crates/bevy_asset/src/loader_builders.rs

```diff
@@ -25,28 +25,164 @@ impl ReaderRef<'_> {
+///   it, but the actual loading won't be completed until later.
+///
+///   Use this if you only need a [`Handle`] or [`UntypedHandle`].
+///
+/// - [`Immediate`]: the load request will load the asset right then and there,
+///   waiting until the asset is fully loaded and giving you access to it.
+///
+///   Note that this requires you to `await` a future, so you must be in an
+///   async context to use direct loading. In an asset loader, you will be in
+///   an async context.
+///
+///   Use this if you need the *value* of another asset in order to load the
+///   current asset. For example, if you are deriving a new asset from the
+///   referenced asset, or you are building a collection of assets. This will
+///   add the path of the asset as a "load dependency".
+///
+///   If the current loader is used in a [`Process`] "asset preprocessor",
+///   such as a [`LoadTransformAndSave`] preprocessor, changing a "load
+///   dependency" will result in re-processing of the asset.
+///
+/// # Load kickoff
+///
+/// If the current context is a normal [`AssetServer::load`], an actual asset
+/// load will be kicked off immediately, which ensures the load happens as soon
+/// as possible. "Normal loads" kicked from within a normal Bevy App will
+/// generally configure the context to kick off loads immediately.
+///
+/// If the current context is configured to not load dependencies automatically
+/// (ex: [`AssetProcessor`]), a load will not be kicked off automatically. It is
+/// then the calling context's responsibility to begin a load if necessary.
 ///
 /// # Lifetimes
+///
 /// - `ctx`: the lifetime of the associated [`AssetServer`](crate::AssetServer) reference
 /// - `builder`: the lifetime of the temporary builder structs
-pub struct NestedLoader<'ctx, 'builder> {
+///
+/// [`direct`]: Self::direct
+/// [`indirect`]: Self::indirect
+/// [type name]: core::any::type_name
+/// [`load`]: Self::load
+/// [`erased`]: Self::erased
+/// [`AssetServer::load`]: crate::AssetServer::load
+/// [`AssetProcessor`]: crate::processor::AssetProcessor
+/// [`Process`]: crate::processor::Process
+/// [`LoadTransformAndSave`]: crate::processor::LoadTransformAndSave
+pub struct NestedLoader<'ctx, 'builder, T, M> {
     load_context: &'builder mut LoadContext<'ctx>,
     meta_transform: Option<MetaTransform>,
-    asset_type_id: Option<TypeId>,
+    typing: T,
+    mode: M,
+}
+
+mod sealed {
+    pub trait Typing {}
+
+    pub trait Mode {}
+}
+
+/// [`NestedLoader`] will be provided the type of asset as a type parameter on
+/// [`load`].
+///
+/// [`load`]: Self::load
+pub struct Typed {
+    _priv: (),
 }
 
-impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
-    pub(crate) fn new(
-        load_context: &'builder mut LoadContext<'ctx>,
-    ) -> NestedLoader<'ctx, 'builder> {
+impl sealed::Typing for Typed {}
+
+/// [`NestedLoader`] has been configured with info on what type of asset to load
+/// at runtime.
+pub struct Erased<'builder> {
```

Comment:
As I stated in the Discord, I would prefer to just call this DynamicType, since erased is a pretty technical term IMO.

But ideally I'd like someone else's opinion since mine is weakly held.

Reply:
After some thought, I agree with `DynamicType` more. It's a bit less concise than `Erased`, but it does match the semantics of `dyn`.

I'll rename `Typed` to `StaticTyped`, `Erased` to `DynamicTyped`, and `UnknownType` to `UnknownTyped`.